proposals/manual-bundling.md

```diff
@@ -0,0 +1,150 @@
+
+   In addition there is a `cordova_plugins.js`, which maps the plugins to a file location.
+
+   Also there is a `plugins/cordova-plugin-camera/www/*`, which contains the files from https://github.com/apache/cordova-plugin-camera/tree/master/www, but with a `.` as a path separator, whereas in any other situation this would be a `/`.
+
+   However, if I take a look at what the `cordova` package on npm is, it's the CLI, which is totally unrelated to the UMD bundle.
+   
+   In addition to `/platforms`, there's also a `/plugins` directory. This contains the following:
+   
+   - Copies of the plugins that already exist in `node_modules`, but with the dotfiles missing.
+   - `android.json` — This appears to duplicate some info from `package.json` and `config.xml`, but in another format.
+   - `fetch.json` — This appears to duplicate some info from `package.json`, but in yet another format.
+
+
+## Proposal
+
+As a modern JavaScript developer, I would like to use *<this year's popular tool>* to bundle my dependencies and code.
+
+- Get rid of `cordova.js`.
+- Get rid of `www/`.
+- Get rid of `plugins/`.
+
+### 1. Load plugins from the `node_modules` directory, not `plugins`
+
+Cordova already downloads packages to `node_modules` using `cordova-fetch`. There's no need to maintain a copy of them in the `plugins` directory.
+
+This also allows plugin creators to use npm dependencies for both hooks and browser code the way this would be done for other packages.
+
+### 2. Let a developer require Cordova using its own build tools.
+
+Based on [cordova-plugin-camera] as an example:
+
+As a developer, I'd like to turn this:
+
+```js
+navigator.camera.getPicture(cameraSuccess, cameraError, cameraOptions)
+```
+
+Into this:
+
+```js
+import CordovaCamera from 'cordova-plugin-camera';
+
+
+CordovaCamera.getPicture(cameraSuccess, cameraError, cameraOptions);
+```
+
+This can be done by adding the following to `package.json`:
+
+```json
+  "main": "www/Camera.js"
+```
+
+### 3. Use [cordova-js] as a proxy to the platform
+
+Instead of using a custom merge of [cordova-js] and *cordova-`<platform name>`*, the Cordova platform would register itself at runtime, so *cordova-js* basically works as a proxy to the registered platform. For example:
+
+In [cordova-android] `package.json`:
+
+```json
+  "main": "cordova-src-js/index.js"
+```
+
+In [cordova-android] `index.js`:
+
+```js
+const cordova = require('cordova-js');
+
+const platform = require('./platform');
+
+
+cordova.setPlatform(platform);
+```
+
+In the Cordova app code:
+
+```js
+// CORDOVA_PLATFORM is set by the build tool, e.g. using webpack.DefinePlugin.
+require(CORDOVA_PLATFORM)
+```
+
+Plugins would make calls to `cordova-js/exec` (replacing `cordova/exec`) independant of the platform. If a developer wishes to use an outdated plugin, it's the developer's responsibility to alias `cordova/*` calls to `cordova-js/*`.
+
+
+### 4. Dependency management
```

Comment:
> We looked into peer dependencies at one point. We found out they were getting deprecated. https://docs.npmjs.com/files/package.json#peerdependencies

> They don't auto install peer deps anymore. They just omit warnings to the user.

> The reason why cordova needs to manage plugin deps is because we can't allow a plugin to have two different versions installed in one project.

> For example, say file-transfer plugin requires file version 3. And say media plugin requires file version 4. This would and should cause an issue. Since the native implementations of file would clash on the native side (duplication errors).

> npm has never given us a way to manage this unique case so we have always had to have our own level of dep management


Their behaviour has changed between npm versions 2 and 3, but it's not deprecated as far as I know.

I believe giving a warning **is** the way npm has given to manage flat dependencies.

- ESLint recommends it for shareable configs: http://eslint.org/docs/developer-guide/shareable-configs#publishing-a-shareable-config
- React libraries add `react` as a peer dependency.
- Angular even uses peer dependencies for depending on `@angular/*` packages.

I don't think the situation for Cordova is any different. How npm should handle these missing dependencies is out of scope for Cordova. Users shouldn't be surprised if their builds fail when using mismatched versions.